src/frame.cpp

```diff
@@ -0,0 +1,371 @@
+                    && abs(vertices[a].z - vertices[b].z) < threshold && abs(vertices[a].z - vertices[c].z) < threshold
+                    && abs(vertices[b].z - vertices[d].z) < threshold && abs(vertices[c].z - vertices[d].z) < threshold)
+                {
+                    if (index2reducedIndex.count(a) == 0 || index2reducedIndex.count(b) == 0 || index2reducedIndex.count(c) == 0 ||
+                        index2reducedIndex.count(d) == 0)
+                        continue;
+
+                    faces.emplace_back(index2reducedIndex[a], index2reducedIndex[d], index2reducedIndex[b]);
+                    faces.emplace_back(index2reducedIndex[d], index2reducedIndex[a], index2reducedIndex[c]);
+                }
+            }
+        }
+
+        std::ofstream out(fname);
+        out << "ply\n";
+        out << "format binary_little_endian 1.0\n";
+        out << "comment pointcloud saved from Realsense Viewer\n";
+        out << "element vertex " << new_vertices.size() << "\n";
+        out << "property float" << sizeof(float) * 8 << " x\n";
+        out << "property float" << sizeof(float) * 8 << " y\n";
+        out << "property float" << sizeof(float) * 8 << " z\n";
+        if (texture)
+        {
+            out << "property uchar red\n";
+            out << "property uchar green\n";
+            out << "property uchar blue\n";
+        }
+        out << "element face " << faces.size() << "\n";
+        out << "property list uchar int vertex_indices\n";
+        out << "end_header\n";
+        out.close();
+
+        out.open(fname, std::ios_base::app | std::ios_base::binary);
+        for (int i = 0; i < new_vertices.size(); ++i)
+        {
+            // we assume little endian architecture on your device
+            out.write(reinterpret_cast<const char*>(&(new_vertices[i].x)), sizeof(float));
+            out.write(reinterpret_cast<const char*>(&(new_vertices[i].y)), sizeof(float));
+            out.write(reinterpret_cast<const char*>(&(new_vertices[i].z)), sizeof(float));
+
+            if (texture)
+            {
+                uint8_t x, y, z;
+                std::tie(x, y, z) = new_tex[i];
+                out.write(reinterpret_cast<const char*>(&x), sizeof(uint8_t));
+                out.write(reinterpret_cast<const char*>(&y), sizeof(uint8_t));
+                out.write(reinterpret_cast<const char*>(&z), sizeof(uint8_t));
+            }
+        }
+        auto size = faces.size();
+        for (int i = 0; i < size; ++i) {
+            int three = 3;
+            out.write(reinterpret_cast<const char*>(&three), sizeof(uint8_t));
+            out.write(reinterpret_cast<const char*>(&(std::get<0>(faces[i]))), sizeof(int));
+            out.write(reinterpret_cast<const char*>(&(std::get<1>(faces[i]))), sizeof(int));
+            out.write(reinterpret_cast<const char*>(&(std::get<2>(faces[i]))), sizeof(int));
+        }
+    }
+
+    size_t points::get_vertex_count() const
```

Comment:
Can we separate points into its own .h/.cpp?

Reply:
👍